schnorr_fun/src/share_backup.rs

```diff
@@ -0,0 +1,319 @@
+    /// Create a new share backup.
+    ///
+    /// ## Panics
+    ///
+    /// The threshold must be greater than 0 and less than 1024.
+    /// The secret share is checked to confirm that its image lies on the public point polynomial.
+    pub fn new<H: Default + Digest<OutputSize = U32>>(
+        polynomial: &Vec<Point>,
+        secret_share: &Scalar,
+        share_index: &Scalar<Public>,
+    ) -> Self {
+        let threshold = polynomial.len() as u16;
+        let identifier = polynomial_identifier::<H>(polynomial);
+
+        if threshold == 0 {
+            panic!("Polynomial threshold can not be zero");
+        }
+        if threshold > 1024 {
+            panic!("Polynomial has too high of a threshold, {threshold} > 1024");
+        }
+
+        if poly::point::eval(polynomial, *share_index) != g!(secret_share * G) {
+            panic!("Secret share is not valid with respect to the polynomial!")
+        }
+
+        Self {
+            threshold,
+            identifier,
+            secret_share: *secret_share,
+            share_index: *share_index,
+        }
+    }
+}
+
+impl fmt::Display for ShareBackup {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut data = [u5::default(); 2 + 4 + 52 + 52];
+
+        let threshold_u5 = (self.threshold - 1).to_le_bytes().to_vec().to_base32();
+        for (i, byte) in threshold_u5.into_iter().take(2).enumerate() {
+            data[i] = byte;
+        }
+
+        for (i, byte) in self.identifier.into_iter().enumerate() {
+            data[2 + i] = byte;
+        }
+
+        let secret_share_u5 = self.secret_share.to_bytes().to_vec().to_base32();
+        for (i, byte) in secret_share_u5.into_iter().enumerate() {
+            data[2 + 4 + i] = byte;
+        }
+
+        let is_small = self.share_index.to_bytes()[0..31].iter().all(|b| *b == 0)
+            && self.share_index.to_bytes()[31] < 32;
+
+        let n_unused_bytes = if is_small {
+            let share_index_u5 =
+                u5::try_from_u8(self.share_index.to_bytes()[31]).expect("must be less than 32");
+            data[2 + 4 + 52] = share_index_u5;
+            52 - 1
+        } else {
+            let share_index_u5 = self.share_index.to_bytes().to_base32();
+            for (i, byte) in share_index_u5.iter().enumerate() {
+                data[2 + 4 + 52 + i] = *byte;
+            }
+            0
+        };
+
+        bech32::encode_to_fmt(f, "frost", &data[..(data.len() - n_unused_bytes)], Bech32m)
+            .expect("hrp is valid and is standard")
+    }
+}
+
+/// Encode a share backup into base32
+///
+/// Use [`ShareBackup`] to securely create secret share backups that are valid w.r.t the polynomial.
+pub fn encode_backup(
+    threshold: u16,
+    identifier: [u5; 4],
+    secret_share: Scalar,
+    share_index: Scalar<Public>,
+) -> String {
+    if threshold == 0 {
+        panic!("Polynomial threshold can not be zero");
+    }
+    if threshold > 1024 {
```

Comment:
These panics are duplicated in `ShareBackup::new` can we design this so they are only done in one spot? I think this function can probably be deleted now that we have `ShareBackup`.

Also it's better to use `assert!` and `assert_ne!` here because they give slightly nicer errors with less effort.